File: sdk/src/main/java/com/sygic/travel/sdk/StSDK.java
package com.sygic.travel.sdk;

import android.content.Context;

import com.sygic.travel.sdk.contentProvider.api.Callback;
import com.sygic.travel.sdk.contentProvider.api.StApi;
import com.sygic.travel.sdk.contentProvider.api.StApiGenerator;
import com.sygic.travel.sdk.contentProvider.api.StObserver;
import com.sygic.travel.sdk.model.api.MediaResponse;
import com.sygic.travel.sdk.model.api.PlaceDetailedResponse;
import com.sygic.travel.sdk.model.api.PlacesResponse;
import com.sygic.travel.sdk.model.media.Medium;
import com.sygic.travel.sdk.model.place.Place;
import com.sygic.travel.sdk.model.query.Query;

import java.io.File;
import java.util.List;

import retrofit2.adapter.rxjava.Result;
import rx.Observable;
import rx.Scheduler;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * <p>Provides public methods for requesting API.</p>
 */
public class StSDK {

	/*-------------------------------------------
	    		       CONSTANTS
	 -------------------------------------------*/

	private static final String TAG = StSDK.class.getSimpleName();

	/*-------------------------------------------
	    		PUBLIC MEMBERS & METHODS
	 -------------------------------------------*/

	public static volatile StSDK instance = null;

	/**
	 * @return Instance of {@link StSDK}.
	 */
	public static StSDK getInstance() {
		if (instance == null) {
			synchronized(StSDK.class) {
				if (instance == null) {
					instance = new StSDK();
				}
			}
		}
		return instance;
	}

	/**
	 * Creates and sends a request to get places, e.g. for map or list.
	 * @param query Query encapsulating data for API request.
	 * @param back Callback. Either {@link Callback#onSuccess(Object)} with places is called, or
	 *             {@link Callback#onFailure(Throwable)} in case of an error is called.
	 */
	public void getPlaces(
		Query query,
		Callback<List<Place>> back
	){
		Observable<Result<PlacesResponse>> unpreparedObservable = getStApi().getPlaces(
			query.getQuery(),
			query.getLevelsQueryString(),
			query.getCategoriesQueryString(),
			query.getMapTilesQueryString(),
			query.getMapSpread(),
			query.getBoundsQueryString(),
			query.getTagsQueryString(),
			query.getParentsQueryString(),
			query.getLimit()
		);
		Observable<Result<PlacesResponse>> preparedObservable = getPreparedObservable(unpreparedObservable);
		subscription = preparedObservable.subscribe(new StObserver<PlacesResponse>(back, false));
	}

	/**
	 * <p>Creates and sends a request to get place with detailed information.</p>
	 * @param id Unique id of a place - detailed information about this place will be requested.
	 * @param back Callback. Either {@link Callback#onSuccess(Object)} with places is called, or
	 *             {@link Callback#onFailure(Throwable)} in case of an error is called.
	 */
	public void getPlaceDetailed(String id, Callback<Place> back){
		Observable<Result<PlaceDetailedResponse>> unpreparedObservable = getStApi().getPlaceDetailed(id);
		Observable<Result<PlaceDetailedResponse>> preparedObservable = getPreparedObservable(unpreparedObservable);
		subscription = preparedObservable.subscribe(new StObserver<PlaceDetailedResponse>(back, false));
	}

	/**
	 * <p>Creates and sends a request to get the place's media.</p>
	 * @param id Unique id of a place - media for this place will be requested.
	 * @param back Callback. Either {@link Callback#onSuccess(Object)} with places is called, or
	 *             {@link Callback#onFailure(Throwable)} in case of an error is called.
	 */
	public void getPlaceMedia(String id, Callback<List<Medium>> back){
		Observable<Result<MediaResponse>> unpreparedObservable = getStApi().getPlaceMedia(id);
		Observable<Result<MediaResponse>> preparedObservable = getPreparedObservable(unpreparedObservable);
		subscription = preparedObservable.subscribe(new StObserver<MediaResponse>(back, false));
	}

	/*-------------------------------------------
	    		          RX
	 -------------------------------------------*/

	/**
	 * <p>Prepares an {@link Observable} - sets {@link Scheduler schedulers}.</p>
	 *
	 * @param unpreparedObservable Observable to be prepared.
	 * @return Observable ready to be subscribed to.
	 */
	public static <T> Observable<T> getPreparedObservable(
		Observable<T> unpreparedObservable
	){
		return unpreparedObservable
			.subscribeOn(Schedulers.newThread())
			.observeOn(AndroidSchedulers.mainThread());
	}

	/**
	 * <p>Unsubscribes a subscribed observable.</p>
	 */
	public void unsubscribeObservable(){
		if(subscription != null && !subscription.isUnsubscribed()){
			subscription.unsubscribe();
		}
	}

	/*-------------------------------------------
	    		PRIVATE MEMBERS & METHODS
	 -------------------------------------------*/

	private StApi stApi;
	private File cacheDir;
	private Subscription subscription;

	private StSDK() {
	}

	/**
	 * <p>Initialization of the SDK.</p>
	 */
	public static void initialize(String xApiKey, Context context) {
		StApiGenerator.headersInterceptor.updateXApiKey(xApiKey);
		StSDK.getInstance().setCacheDir(context.getCacheDir());
	}

	/**
	 * <p>Creates and returns an instance of the {@link StApi} interface.</p>
	 * @return Instance of the {@link StApi} interface.
	 */
	private StApi getStApi() {
		if(stApi == null){
			stApi = StApiGenerator.createStApi(StApi.class, cacheDir);
		}
		return stApi;
	}

	/**
	 * <p>Sets a cache directory.</p>
	 * @param cacheDir Cache directorry.
	 */
	private void setCacheDir(File cacheDir) {
		this.cacheDir = cacheDir;
	}
}

File: sdk/src/main/java/com/sygic/travel/sdk/StSDK.kt
package com.sygic.travel.sdk

import android.arch.persistence.room.Room
import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import com.sygic.travel.sdk.api.StApi
import com.sygic.travel.sdk.api.StApiGenerator
import com.sygic.travel.sdk.db.StDb
import com.sygic.travel.sdk.model.media.Medium
import com.sygic.travel.sdk.model.place.Place
import com.sygic.travel.sdk.model.place.Tour
import com.sygic.travel.sdk.model.query.PlacesQuery
import com.sygic.travel.sdk.model.query.ToursQuery
import com.sygic.travel.sdk.provider.DataProvider
import com.sygic.travel.sdk.utils.runWithCallback
import java.io.File

/**
 * Provides public methods for requesting API.
 */
class StSDK internal constructor(context: Context, cacheDir: File) {
	companion object {
		private val ST_SDK_NAME_AND_VERSION = "sygic-travel-sdk-android/sdk-version"
		private val ANDROID_VERSION = "Android/" + Build.VERSION.RELEASE
		private val DATABASE_NAME = "st-sdk-db"

		/**
		 * Creates the SDK.
		 * @param xApiKey Api key - must be provided.
		 * @param context Application's context.
		 */
		fun create(xApiKey: String, context: Context): StSDK {
			StApiGenerator.headersInterceptor.setApiKey(xApiKey)
			StApiGenerator.headersInterceptor.setUserAgent(createUserAgent(context))
			return StSDK(context, context.cacheDir)
		}

		private fun createUserAgent(context: Context): String {
			var packageInfo: PackageInfo? = null
			try {
				packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
			} catch (e: PackageManager.NameNotFoundException) {
				e.printStackTrace()
			}

			var userAgent = ""
			userAgent += context.packageName + "/"
			userAgent += if (packageInfo != null) packageInfo.versionName + " " else ""
			userAgent += ST_SDK_NAME_AND_VERSION + " "
			userAgent += ANDROID_VERSION
			return userAgent
		}
	}


	private var stApi: StApi = StApiGenerator.createStApi(StApi::class.java, cacheDir)
	private var stDb: StDb = Room.databaseBuilder(context, StDb::class.java, DATABASE_NAME).build()
	private var dataProvider: DataProvider = DataProvider(stApi, stDb)


	/**
	 * Creates and sends a request to get places, e.g. for map or list.
	 * @param placesQuery PlacesQuery encapsulating data for API request.
	 */
	fun getPlaces(placesQuery: PlacesQuery, callback: Callback<List<Place>?>) {
		runWithCallback({ dataProvider.getPlaces(placesQuery) }, callback)
	}


	/**
	 * Creates and sends a request to get place with detailed information.
	 * @param id Unique id of a place - detailed information about this place will be requested.
	 */
	fun getPlaceDetailed(id: String, callback: Callback<Place?>) {
		runWithCallback({ dataProvider.getPlaceDetailed(id) }, callback)
	}


	/**
	 * Creates and sends a request to get places with detailed information.
	 * @param ids Ids of places - detailed information about these places will be requested.
	 */
	fun getPlacesDetailed(ids: List<String>, callback: Callback<List<Place>?>) {
		runWithCallback({ dataProvider.getPlacesDetailed(ids) }, callback)
	}


	/**
	 * Creates and sends a request to get the place's media.
	 * @param id Unique id of a place - media for this place will be requested.
	 */
	fun getPlaceMedia(id: String, callback: Callback<List<Medium>?>) {
		runWithCallback({ dataProvider.getPlaceMedia(id) }, callback)
	}


	/**
	 * Creates and sends a request to get the Tours.
	 * @param toursQuery ToursQuery encapsulating data for API request.
	 */
	fun getTours(toursQuery: ToursQuery, callback: Callback<List<Tour>?>) {
		runWithCallback({ dataProvider.getTours(toursQuery) }, callback)
	}


	/**
	 * Stores a place's id in a local persistent storage. The place is added to the favorites.
	 * @param id A place's id, which is stored.
	 */
	fun addPlaceToFavorites(id: String, callback: Callback<Unit>) {
		runWithCallback({ dataProvider.addPlaceToFavorites(id) }, callback)
	}


	/**
	 * Removes a place's id from a local persistent storage. The place is removed from the favorites.
	 * @param id A place's id, which is removed.
	 */
	fun removePlaceFromFavorites(id: String, callback: Callback<Unit>) {
		runWithCallback({ dataProvider.removePlaceFromFavorites(id) }, callback)
	}


	/**
	 * Method returns a list of all favorite places' ids.
	 */
	fun getFavoritesIds(callback: Callback<List<String>?>) {
		runWithCallback({ dataProvider.getFavoritesIds() }, callback)
	}
}
